File: app/src/main/java/com/etesync/syncadapter/ui/RemoveMemberFragment.java
package com.etesync.syncadapter.ui;

import android.accounts.Account;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;

import com.etesync.syncadapter.AccountSettings;
import com.etesync.syncadapter.Constants;
import com.etesync.syncadapter.HttpClient;
import com.etesync.syncadapter.InvalidAccountException;
import com.etesync.syncadapter.R;
import com.etesync.syncadapter.journalmanager.JournalManager;
import com.etesync.syncadapter.model.CollectionInfo;

import org.apache.commons.codec.Charsets;

import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;

public class RemoveMemberFragment extends DialogFragment {
    final static private String KEY_MEMBER = "memberEmail";
    private AccountSettings settings;
    private OkHttpClient httpClient;
    private HttpUrl remote;
    private CollectionInfo info;
    private String memberEmail;

    public static RemoveMemberFragment newInstance(Account account, CollectionInfo info, String email) {
        RemoveMemberFragment frag = new RemoveMemberFragment();
        Bundle args = new Bundle(1);
        args.putParcelable(Constants.KEY_ACCOUNT, account);
        args.putSerializable(Constants.KEY_COLLECTION_INFO, info);
        args.putString(KEY_MEMBER, email);
        frag.setArguments(args);
        return frag;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Account account = getArguments().getParcelable(Constants.KEY_ACCOUNT);
        info = (CollectionInfo) getArguments().getSerializable(Constants.KEY_COLLECTION_INFO);
        memberEmail = getArguments().getString(KEY_MEMBER);
        try {
            settings = new AccountSettings(getContext(), account);
            httpClient = HttpClient.create(getContext(), settings);
        } catch (InvalidAccountException e) {
            e.printStackTrace();
        }
        remote = HttpUrl.get(settings.getUri());

        new MemberRemove().execute();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        ProgressDialog progress = new ProgressDialog(getContext());
        progress.setTitle(R.string.collection_members_removing);
        progress.setMessage(getString(R.string.please_wait));
        progress.setIndeterminate(true);
        progress.setCanceledOnTouchOutside(false);
        setCancelable(false);
        return progress;
    }

    private class MemberRemove extends AsyncTask<Void, Void, MemberRemove.RemoveResult> {
        @Override
        protected RemoveResult doInBackground(Void... voids) {
            try {
                JournalManager journalsManager = new JournalManager(httpClient, remote);
                JournalManager.Journal journal = JournalManager.Journal.fakeWithUid(info.uid);

                JournalManager.Member member = new JournalManager.Member(memberEmail, "placeholder".getBytes(Charsets.UTF_8));
                journalsManager.deleteMember(journal, member);

                return new RemoveResult(null);
            } catch (Exception e) {
                return new RemoveResult(e);
            }
        }

        @Override
        protected void onPostExecute(RemoveResult result) {
            if (result.throwable == null) {
                ((Refreshable) getActivity()).refresh();
            } else {
                new AlertDialog.Builder(getActivity())
                        .setIcon(R.drawable.ic_error_dark)
                        .setTitle(R.string.collection_members_remove_error)
                        .setMessage(result.throwable.getMessage())
                        .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {

                            }
                        }).show();
            }
            dismiss();
        }

        class RemoveResult {
            final Throwable throwable;

            RemoveResult(final Throwable throwable) {
                this.throwable = throwable;
            }
        }
    }
}

File: app/src/main/java/com/etesync/syncadapter/ui/RemoveMemberFragment.kt
package com.etesync.syncadapter.ui

import android.accounts.Account
import android.app.Dialog
import android.app.ProgressDialog
import android.os.AsyncTask
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import com.etesync.syncadapter.*
import com.etesync.syncadapter.journalmanager.JournalManager
import com.etesync.syncadapter.model.CollectionInfo
import okhttp3.HttpUrl

class RemoveMemberFragment : DialogFragment() {
    private var settings: AccountSettings? = null
    private var info: CollectionInfo? = null
    private var memberEmail: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val account = arguments!!.getParcelable<Account>(Constants.KEY_ACCOUNT)
        info = arguments!!.getSerializable(Constants.KEY_COLLECTION_INFO) as CollectionInfo
        memberEmail = arguments!!.getString(KEY_MEMBER)
        try {
            settings = AccountSettings(context!!, account!!)
        } catch (e: InvalidAccountException) {
            e.printStackTrace()
        }

        MemberRemove().execute()
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val progress = ProgressDialog(context)
        progress.setTitle(R.string.collection_members_removing)
        progress.setMessage(getString(R.string.please_wait))
        progress.isIndeterminate = true
        progress.setCanceledOnTouchOutside(false)
        isCancelable = false
        return progress
    }

    private inner class MemberRemove : AsyncTask<Void, Void, MemberRemove.RemoveResult>() {
        override fun doInBackground(vararg voids: Void): RemoveResult {
            val httpClient = HttpClient.Builder(context, settings).build()

            try {
                val remote = HttpUrl.get(settings!!.uri!!)

                val journalsManager = JournalManager(httpClient.okHttpClient, remote!!)
                val journal = JournalManager.Journal.fakeWithUid(info!!.uid!!)

                val member = JournalManager.Member(memberEmail!!, "placeholder".toByteArray())
                journalsManager.deleteMember(journal, member)

                return RemoveResult(null)
            } catch (e: Exception) {
                return RemoveResult(e)
            } finally {
                httpClient.close()
            }
        }

        override fun onPostExecute(result: RemoveResult) {
            if (result.throwable == null) {
                (activity as Refreshable).refresh()
            } else {
                AlertDialog.Builder(activity!!)
                        .setIcon(R.drawable.ic_error_dark)
                        .setTitle(R.string.collection_members_remove_error)
                        .setMessage(result.throwable.message)
                        .setPositiveButton(android.R.string.yes) { _, _ -> }.show()
            }
            dismiss()
        }

        internal inner class RemoveResult(val throwable: Throwable?)
    }

    companion object {
        private val KEY_MEMBER = "memberEmail"

        fun newInstance(account: Account, info: CollectionInfo, email: String): RemoveMemberFragment {
            val frag = RemoveMemberFragment()
            val args = Bundle(1)
            args.putParcelable(Constants.KEY_ACCOUNT, account)
            args.putSerializable(Constants.KEY_COLLECTION_INFO, info)
            args.putString(KEY_MEMBER, email)
            frag.arguments = args
            return frag
        }
    }
}
